pubnub-core/src/subscription/subscribe_loop_supervisor.rs

```diff
@@ -0,0 +1,161 @@
+use super::registry::Registry;
+use super::subscribe_loop::{
+    subscribe_loop, ControlCommand, ControlTx, ExitTx, ListenerType, SubscribeLoopParams,
+};
+use super::subscription::Subscription;
+use crate::runtime::Runtime;
+use crate::transport::Transport;
+use crate::PubNub;
+use futures_channel::{mpsc, oneshot};
+use futures_util::sink::SinkExt;
+use log::debug;
+
+/// SubscribeLoopSupervisor is responsible for the lifecycle of the subscribe
+/// loop.
+/// It owns the subscribe loop control handle and provides a high-level
+/// interface to the control operations.
+/// It will intelligently spawn and respawn the subscribe loop on a need basis.
+///
+/// Deliberately doesn't implement `Clone` to avoid issues with improper
+/// duplication of control handles.
+#[derive(Debug)]
+pub(crate) struct SubscribeLoopSupervisor {
+    /// Configuration params.
+    params: SubscribeLoopSupervisorParams,
+
+    /// Control handle to the subscribe loop.
+    control_tx: Option<ControlTx>,
+}
+
+/// SubscribeLoopSupervisorParams configuration params.
+#[derive(Debug)]
+pub(crate) struct SubscribeLoopSupervisorParams {
+    /// If set, gets a signal when subscribe loop exits.
+    pub exit_tx: Option<ExitTx>,
+}
+
+impl SubscribeLoopSupervisor {
+    pub fn new(params: SubscribeLoopSupervisorParams) -> Self {
+        Self {
+            params,
+            control_tx: None,
+        }
+    }
+}
+
+impl SubscribeLoopSupervisor {
+    pub async fn subscribe<'a, TTransport, TRuntime>(
+        &mut self,
+        pubnub: &'a mut PubNub<TTransport, TRuntime>,
+        channel: &str,
+    ) -> Subscription<TRuntime>
+    where
+        TTransport: Transport + 'static,
+        TRuntime: Runtime + 'static,
+    {
+        // Since recursion is troublesome with async fns, we use the loop trick.
+        let (id, control_tx, channel_rx) = loop {
+            let (channel_tx, channel_rx) = mpsc::channel(10);
+
+            let id_or_retry = if let Some(ref mut control_tx) = self.control_tx {
+                // Send a command to add the channel to the running
+                // subscribe loop.
+
+                debug!("Adding channel {:?} to the running loop", channel);
+
+                let (id_tx, id_rx) = oneshot::channel();
+
+                // TODO: unify interfaces to either use `ListenerType` or
+                // `&str` when we refer to a channel.
```

Comment:
Yes. This method needs to be able to create a `Subscription` for either a channel or a channel-group.

Reply:
As I see, the channel/channel-group abstraction was not implemented to completion when I started, otherwise, it'd be integrated globally and I wouldn't be problematic here.
I don't want to solve this within this PR, but it can definitely get some polishing later on.

Reply:
The incompleteness was partially my fault. Not making excuses, but I just didn't need the channel groups abstraction at the time.

Fair, we don't have to make any unrelated improvements in this PR. I was only commenting on the new comment added in the PR.